gdx/src/com/badlogic/gdx/math/Octree.java

```diff
@@ -0,0 +1,307 @@
+            for (OctreeNode child : children) {
+                for (T geometry : geometries) {
+                    child.add(geometry, maxItemsPerNode, collider);
+                }
+            }
+        }
+
+        public boolean contains (Collider<T> boundingBoxHandler, T geometry) {
+            return boundingBoxHandler.intersects(bounds, geometry);
+        }
+
+        public void add (T geometry, int maxItemsPerNode, Collider<T> collider) {
+            // If is not leaf, check children
+            if (!isLeaf()) {
+                for (OctreeNode child : children) {
+                    child.add(geometry, maxItemsPerNode, collider);
+                }
+                return;
+            }
+
+            if (!collider.intersects(bounds, geometry)) {
+                return;
+            }
+
+            addGeometry(geometry);
+            if (geometries.size > maxItemsPerNode && level > 0) {
+                split(maxItemsPerNode, collider);
+            }
+        }
+
+        public void remove (T aabb) {
+            if (isLeaf()) {
+                if (geometries != null) {
+                    geometries.remove(aabb);
+                }
+            } else {
+                for (OctreeNode node : children) {
+                    node.remove(aabb);
+                }
+            }
+        }
+
+        public boolean isLeaf () {
+            return children == null;
+        }
+
+        private void addGeometry (T geometry) {
+            if (geometries == null) {
+                geometries = new ObjectSet<>();
+            }
+            geometries.add(geometry);
+        }
+
+        public ObjectSet<T> query (BoundingBox aabb, ObjectSet<T> result) {
+            if (!isLeaf()) {
+                for (OctreeNode node : children) {
+                    node.query(aabb, result);
+                }
+            }
+            if (aabb.contains(bounds)) {
+                if (geometries != null) {
+                    result.addAll(geometries);
+                }
+            }
+            return result;
+        }
+
+        public ObjectSet<T> query (Frustum frustum, ObjectSet<T> result) {
+            if (!isLeaf()) {
+                for (OctreeNode node : children) {
+                    node.query(frustum, result);
+                }
+            }
+            if (frustum.boundsInFrustum(bounds)) {
+                if (geometries != null) {
+                    result.addAll(geometries);
+                }
+            }
+            return result;
+        }
+
+        Vector3 tmp = new Vector3();
+        public void rayCast (Ray ray, float maxDistance, Collider<T> narrowPhase, RayCastResult<T> result) {
+            // Check intersection with node
+            boolean intersect = Intersector.intersectRayBounds(ray, bounds, tmp);
+            if (!intersect) {
+                return;
+            } else {
+                float dst2 = tmp.dst2(ray.origin);
+                if (dst2 > maxDistance) {
```

Comment:
superior to maxDistance squared

Reply:
Right! What if `maxDistance = Float.MAX_FLOAT;`
Should I set the default maxDistance to `Math.sqrt(Float.MAX_FLOAT)` instead?

Reply:
Nvm, fixed.